### ISSUE TITLE
[Ray Train Feature Request] Native Cross-Validation Support in Ray Train API

### ISSUE THREAD

**Issue by shoaibkhanz** [ray-project/ray#53581]
### Description

Current workflows for e.g. for XGBoostTrainer etc do not allow for cross validation 
- (https://scikit-learn.org/stable/modules/cross_validation.html) 
- https://xgboost.readthedocs.io/en/latest/python/examples/cross_validation.html

The user has to implement cross validation manually using for loops which arent vectorised and thus take a considerable time to complete.

Also in my experiments, scikit learn models that implement cv arent really well parallelised and take longer as well. Thus this feature request suggest that we implement cross validation within Ray Train API, so can be applied across any framework.

*see use case section in this issue for more details*

### Use case

I want to efficiently perform cross-validation at scale, using distributed compute, as part of my end-to-end machine learning workflows on Ray.

**Context:**
Cross-validation is a core part of robust ML model development. It helps estimate the generalisation ability of models and is a standard practice for evaluating algorithms, tuning hyperparameters, and comparing models. Most ML practitioners expect to use cross-validation as a simple, integrated step in their workflow.
- K-Fold Cross-Validation
- Stratified K-Fold Cross-Validation
- Group K-Fold
- Time Series Split (I can see an argument for this due to its complexity and how each framework deal with this, that perhaps frameworks like Nixtla, Darts implement so this can be off loaded to them (https://nixtlaverse.nixtla.io/mlforecast/docs/how-to-guides/cross_validation.html#cross-validation)) 

**Current Situation:**
When working with Ray’s trainers (such as XGBoostTrainer etc.), there is no built-in support for cross-validation.

As a user, this means I need to write my own loops or orchestrate manual splitting and training, which is both time-consuming and less efficient. Standard libraries (e.g., scikit-learn, XGBoost) do provide cross-validation, but they either do not scale to distributed settings or do not fully utilise cluster resources.

In practice, this makes large-scale cross-validation slow and cumbersome, especially with big datasets or more complex models.

**What I want:**
A way to seamlessly and efficiently apply cross-validation when training models with Ray—leveraging Ray’s distributed capabilities—so that I can:

Use cross-validation as a first-class citizen in my training workflow
- Avoid reimplementing common evaluation logic
- Run cross-validation jobs faster and more efficiently, at any scale

**Ideal outcome:**
When I train a model using Ray, I want cross-validation to be as accessible and scalable as the rest of the API—making it easy to get robust model performance estimates, without extra boilerplate or resource underutilisation.

**Comment by EmaFerrao** [comment]
Hi! @pedrojeronim0 and I are interested in picking this up and would love to get some thoughts before we start coding.

## Proposed Design

We’re currently thinking of implementing cross-validation as a lightweight orchestration layer on top of Ray Train, without modifying existing trainer APIs.

Conceptually, this would:

- take a single `ray.data.Dataset` as input  
- perform a simple K-Fold split  
- launch one trainer run per fold (executed in parallel across the cluster via Ray)  
- aggregate metrics across folds (mean/std)

The goal is to avoid requiring users to manually orchestrate folds, while keeping the implementation minimally invasive, compatible with existing trainers, and leveraging Ray’s distributed execution capabilities.

For scope, we’re planning to start with K-Fold only, leaving other strategies (Stratified, Group, Time Series) for future iterations.

## Proposed API

```python
results = cross_validate(
    trainer_factory=lambda: TorchTrainer(
        train_func=train_func,
        scaling_config=ScalingConfig(num_workers=4),
    ),
    dataset=dataset,
    splitter=KFoldSplitter(n_splits=5),
)
```

## Expected Output

The initial version would return a simple structured result like:

```python
{
    "folds": [
        {"fold": 0, "metrics": {...}},
        {"fold": 1, "metrics": {...}},
        {"fold": 2, "metrics": {...}},
    ],
    "mean_metrics": {...},
    "std_metrics": {...},
}
```

## Questions

- Is there any recommended pattern in Ray Data for implementing deterministic dataset splits for cross-validation, or should this be done using lower-level dataset operations such as indexing, shuffling, or filtering?

Any feedback would be appreciated!
